pkg/webhook/validating/podgroup.go

```diff
@@ -0,0 +1,115 @@
+	"encoding/json"
+	"fmt"
+
+	"github.com/golang/glog"
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	kbv1 "github.com/kubernetes-sigs/kube-batch/pkg/apis/scheduling/v1alpha1"
+	"github.com/kubernetes-sigs/kube-batch/pkg/webhook/api"
+	"github.com/kubernetes-sigs/kube-batch/pkg/webhook/util"
+)
+
+var podGroupResource = metav1.GroupVersionResource{
+	Group:    kbv1.GroupName,
+	Version:  kbv1.GroupVersion,
+	Resource: "podgroups",
+}
+
+// PodGroupValidator is used for validating whether we could a new PodGroup
+type PodGroupValidator struct {
+	admitterContext api.AdmitterContext
+}
+
+// NewPodGroupValidtor returns a new PodGroupValidator for validating.
+func NewPodGroupValidtor(ctx api.AdmitterContext) (*PodGroupValidator, error) {
+	return &PodGroupValidator{admitterContext: ctx}, nil
+}
+
+// Admit is used to check whether we could create the PodGroup.
+func (pgv *PodGroupValidator) Admit(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	glog.V(2).Info("Admitting podgroup")
+	if ar.Request.Resource != podGroupResource {
+		err := fmt.Errorf("expect resource to be %s", podGroupResource)
+		glog.Errorf("PodGroupValidator receives unexpected resource %+v", ar.Request)
+		return util.ToAdmissionResponse(err)
+	}
+
+	raw := ar.Request.Object.Raw
+	podGroup := kbv1.PodGroup{}
+	if err := json.Unmarshal(raw, &podGroup); err != nil {
+		glog.Errorf("Failed to unmarshal %s: %v", raw, err)
+		return util.ToAdmissionResponse(err)
+	}
+	reviewResponse := &v1beta1.AdmissionResponse{}
+	reviewResponse.Allowed = true
+	if exceed, err := pgv.exceedPodGroupLimit(&podGroup); err != nil {
+		glog.Errorf("Failed to admit PodGroup %s/%s: %v", podGroup.Namespace, podGroup.Name, err)
+		return util.ToAdmissionResponse(err)
+	} else if exceed {
+		return &v1beta1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: "exceed MaxPodGroupNumber",
+			},
+		}
+	}
+
+	return reviewResponse
+}
+
+// exceedPodGroupLimit returns whether the number of PodGroups in the same queue with podGroup will exceed
+// MaxPodGroupNumber after creating podGroup.
+func (pgv *PodGroupValidator) exceedPodGroupLimit(podGroup *kbv1.PodGroup) (bool, error) {
+	// Get queue
+	queueName := podGroup.Spec.Queue
+	if len(queueName) == 0 {
+		glog.V(4).Infof("No queue specified for PodGroup %s/%s", podGroup.Namespace, podGroup.Name)
+		return false, nil
+	}
+	queue, err := pgv.admitterContext.QueueLister.Get(queueName)
+	if err != nil {
+		return true, err
+	}
+	// If not set, default to infinite.
+	if queue.Spec.MaxPodGroupNumber == 0 {
+		return false, nil
+	}
+	// TODO: If the number of PodGroups is large, we should do some optimization, e.g.using informer and cache
+	// to implement it.
+	max := queue.Spec.MaxPodGroupNumber - 1
+	podGroups, err := pgv.admitterContext.PodGroupLister.List(labels.Everything())
+	if err != nil {
+		return true, err
+	}
+	for _, group := range podGroups {
```

Comment:
As we discussed offline, it's better to add a status to Queue and check it here, similar to Quota.